### PR TITLE
Heavy docker-app loading refactoring

### PR DESCRIPTION
**Pull request by vdemeester**
- introduces a `types` package, moving it from `internal/types`
- introduces a `loader` package, with function to lead each type of
  app (folder, tarball, single-file) — only docker image is still in
  `internal/packager`.
- introduces a small `compose` helper package
- updates some `settings` functions
- split `internal/render` into `render` package and `internal/inspect`

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

### CODE REVIEW

File: internal/packager/split.go [hunk 1]
```diff
@@ -29,20 +37,23 @@ func Split(appname string, outputDir string) error {
...
+func Merge(app *types.App, target io.Writer) error {
+	if len(app.Composes()) > 1 {
+		return errors.New("merge: multiple compose files is not supported yet")
+	}
+	if len(app.Settings()) > 1 {
+		return errors.New("merge: multiple setting files is not supported yet")
+	}
+	for _, data := range [][]byte{
+		app.Metadata(),
+		[]byte("\n---\n"),
```
Comment (silvin-lubecki) [inline internal/packager/split.go]: var separator = []byte("\n---\n")

File: loader/loader.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/pkg/errors"
+)
+
+// LoadFromSingleFile loads a docker app from a single-file format (as a reader)
+func LoadFromSingleFile(path string, r io.Reader, ops ...func(*types.App) error) (*types.App, error) {
+	data, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, errors.Wrap(err, "error reading single-file")
+	}
+	parts := strings.Split(string(data), "\n---")
```
Comment (silvin-lubecki) [inline loader/loader.go]: Why don't you split using "\n---\n" ?

File: loader/loader.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+// LoadFromTarReader loads a docker app from a tarball reader
+func LoadFromTarReader(r io.Reader, ops ...func(*types.App) error) (*types.App, error) {
+	dir, err := ioutil.TempDir("", "load-from-tar")
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot load app from tar")
+	}
+	if err := archive.Untar(r, dir, &archive.TarOptions{
+		NoLchown: true,
+	}); err != nil {
+		return nil, errors.Wrap(err, "cannot load app from tar")
```
Comment (silvin-lubecki) [inline loader/loader.go]: The temp dir is not cleaned here.

File: loader/loader_test.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	defer dir.Remove()
+	app, err := LoadFromDirectory(dir.Path())
+	assert.NilError(t, err)
+	assert.Assert(t, app != nil)
+	assert.Assert(t, is.Equal(app.Path, dir.Path()))
+	assert.Assert(t, is.Len(app.Settings(), 1))
+	assertContentIs(t, app.Settings()[0], settings)
+	assert.Assert(t, is.Len(app.Composes(), 1))
+	assertContentIs(t, app.Composes()[0], yaml)
+	assertContentIs(t, app.Metadata(), metadata)
```
Comment (silvin-lubecki) [inline loader/loader_test.go]: Factorize this checking part with TestLoadFromSingleFile and TestLoadFromTar?

File: types/types.go [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	return func(app *App) error {
+		app.Name = name
+		return nil
+	}
+}
+
+// WithPath sets the original path of the app
+func WithPath(path string) func(*App) error {
+	return func(app *App) error {
+		app.Path = path
```
[the remark this answers is not in the document]
Comment (vdemeester) [reply types/types.go]: Why would we at that point ?

File: types/types.go [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+	}
+}
+
+// WithSettingsFiles adds the specified settings files to the app
+func WithSettingsFiles(files ...string) func(*App) error {
+	return func(app *App) error {
+		for _, file := range files {
+			d, err := ioutil.ReadFile(file)
+			if err != nil {
+				return err
```
[the remark this answers is not in the document]
Comment (vdemeester) [reply types/types.go]: :thinking: maybe :stuck_out_tongue_closed_eyes:

File: internal/packager/split.go [hunk 7]
```diff
@@ -7,20 +7,28 @@ import (
...
+	"github.com/docker/app/types"
+	"github.com/pkg/errors"
 )
 
 // Split converts an app package to the split version
-func Split(appname string, outputDir string) error {
-	err := os.Mkdir(outputDir, 0755)
+func Split(app *types.App, outputDir string) error {
+	if len(app.Composes()) > 1 {
+		return errors.New("merge: multiple compose files is not supported yet")
```
[the remark this answers is not in the document]
Comment (vdemeester) [reply internal/packager/split.go]: Yeah, we should support that at some point, but yeah maybe the `yet` shouldn't be part of the error message

File: loader/loader.go (hunk 3 above)
Comment (vdemeester) [reply loader/loader.go]:
```
		types.WithCleanup(func() {
			os.RemoveAll(dir)
		}),
```

File: types/types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (silvin-lubecki) [reply types/types.go]: To fail early?

File: types/types_test.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	assert.NilError(t, err)
+	assert.Assert(t, is.Len(app.Settings(), 1))
+	assertContentIs(t, app.Settings()[0], "foo=bar")
+	assert.Assert(t, is.Len(app.Composes(), 1))
+	assertContentIs(t, app.Composes()[0], yaml)
+	assertContentIs(t, app.Metadata(), "foo")
+}
+
+func TestNewAppWithOpError(t *testing.T) {
+	_, err := NewApp("any-app", func(_ *App) error { return errors.New("error creating") })
```
[the remark this answers is not in the document]
Comment (silvin-lubecki) [reply types/types_test.go]: Ok it's a nit 😸

**Comment by silvin-lubecki** [comment]
```
Linting...
types/types.go:11:7:warning: exported const SingleFileSeparator should have comment or be unexported (golint)
```

File: types/types.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (vdemeester) [reply types/types.go]: Right, but we're *just* building the type `App` here, we're not loading anything…

File: loader/loader.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+func LoadFromTarReader(r io.Reader, ops ...func(*types.App) error) (*types.App, error) {
+	dir, err := ioutil.TempDir("", "load-from-tar")
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot load app from tar")
+	}
+	if err := archive.Untar(r, dir, &archive.TarOptions{
+		NoLchown: true,
+	}); err != nil {
+		if err := os.RemoveAll(dir); err != nil {
+			return nil, errors.Wrap(err, "cannot remove temporary folder")
```
Comment (silvin-lubecki) [inline loader/loader.go]: This error can be misleading, as it shadows the first error. You should Wrap both the errors.

**Comment by silvin-lubecki** [comment]
😱

**Pull request merged**